### PR TITLE
feat(aws_kinesis_streams): Add configurable partition keys

### LINKED ISSUE #358: Configurable Kinesis partition keys
Currently, Vector generates random partition keys in the `aws_kinesis` sink, but we need to provide the ability for the user to use an event field as the key. This is the same as the `kafka` sink's `key_field` option.

- [ ] Add new `partition_key_field` option that accepts a string for the event field to use.
- [ ] The `partition_key_field` field is optional, the default is to generate a random string (which is what currently happens)
- [ ] If the provided key name does not exist log a rate limited warning.

### PR DESCRIPTION
**Pull request by LucioFranco**
This PR adds support for configurable partition keys. This also adds serde support for template so that it can be used directly in configs.

Closes #358

### CODE REVIEW

File: src/sinks/aws_kinesis_streams.rs [hunk 1]
```diff
@@ -305,7 +320,64 @@ mod integration_tests {
...
+            });
+
+        let pump = sink.send_all(iter_ok(events));
+        rt.block_on(pump).unwrap();
+
+        std::thread::sleep(std::time::Duration::from_secs(1));
+
+        let timestamp = timestamp as f64 / 1000.0;
+        let records = rt
+            .block_on(fetch_records(stream.clone(), timestamp, region))
```
Comment (lukesteensen) [inline src/sinks/aws_kinesis_streams.rs]: I don't see how this is testing that half of the events get written to one partition and half get written to another.

**Comment by LucioFranco** [comment]
Last step for this PR is to generate the docs but that will depend on getting gitbook building again...

**Comment by binarylogic** [comment]
Gitbook doesn't build documentation for PRs. How does this PR depend on Gitbook?

**Comment by LucioFranco** [comment]
@binarylogic I am trying to generate the docs locally so that the generated docs are included in this PR. But the generate command fails since the coercer docs are not on the gitbook page yet due to I believe the failing builds on master. Therefore, to finish this we need to get gitbook working again. Unless we are ok with merging without the generated docs changes.

**Comment by binarylogic** [comment]
That makes sense. Then yes, let’s fix doc builds on master.

File: .metadata.toml [hunk 2]
```diff
@@ -834,13 +834,20 @@ description = "The encoding format used to serialize the events before flushing.
...
 [sinks.aws_kinesis_streams.options.stream_name]
 type = "string"
 examples = ["my-stream"]
 null = false
-description = "The [stream name][url.aws_cw_logs_stream_name] of the target CloudWatch Logs stream."
+description = "The [stream name][url.aws_cw_logs_stream_name] of the target Kinesis Logs stream."
+
+[sinks.aws_kinesis_streams.options.partition_key_field]
+type = "string"
+default = "random"
```
Comment (binarylogic) [inline .metadata.toml]: `random` isn't the default here, there is no default so I would remove this.

**Comment by binarylogic** [comment]
I just realized that kinesis only accepts values between 1 and 256 characters. Can we strip values longer than this?

**Comment by binarylogic** [comment]
And how does this work for non-string values? Primitive types are each to coerce, but I'm curious how we handle `timestamp` types.

**Comment by LucioFranco** [comment]
@binarylogic yeah, we can reduce the string to less than 256. For all other types we attempt to convert them all to strings. So it will use the string representation as defined within the `src/event/mod.rs`.

**Pull request merged**